src/psyclone/nemo.py

```diff
@@ -141,33 +151,39 @@ class NemoInvoke(Invoke):
     :type ast: :py:class:`fparser.two.Fortran2003.Main_Program` or \
                :py:class:`fparser.two.Fortran2003.Module`
     :param str name: the name of this Invoke (program unit).
+    :param invokes: the Invokes object that holds this Invoke.
+    :type invokes: :py:class:`psyclone.psyGen.NemoInvokes`
+
     '''
-    def __init__(self, ast, name):
+    def __init__(self, ast, name, invokes):
         # pylint: disable=super-init-not-called
+        from fparser.two.Fortran2003 import Execution_Part, Specification_Part
+        self._invokes = invokes
         self._schedule = None
         self._name = name
         # Store the whole fparser2 AST
         self._ast = ast
-        # A temporary workaround for the fact that we don't yet have a
-        # symbol table to store information on the variable declarations.
-        # TODO (#500) remove this workaround.
-        self._loop_vars = []
-        from fparser.two.Fortran2003 import Execution_Part, Specification_Part
 
         # Find the section of the tree containing the execution part
         # of the code
         exe_part = get_child(ast, Execution_Part)
         if not exe_part:
             # This subroutine has no execution part so we skip it
-            # TODO log this event
+            # TODO #11 log this event
             return
 
         # Store the root of this routine's specification in the AST
         self._spec_part = get_child(ast, Specification_Part)
 
-        # We now walk through the AST produced by fparser2 and construct a
-        # new AST using objects from the nemo module.
-        self._schedule = NemoInvokeSchedule(self, exe_part)
+        # We now walk through the fparser2 parse tree and construct the
+        # PSyIR with a NemoInvokeSchedule at its root.
```

Comment:
Oh yes, that's much nicer.

Reply:
The exe_part tests seems to have been removed completely. I was suggesting that it was moved after NemoFparser2Reader and that we looked at the resultant PSyIR tree to perform the check (rather than the Fortran parse tree). Have I missed something?

Reply:
I did it this way since it's cleaner. If there's no executable part we just get an empty Schedule which feels like the right way to go.